File: data-structures/arrays/2D Array - DS/solution.go
// NOTE to run it
// go run solution.go < input.txt

package main

import (
	"fmt"
	"math"
)

func main() {
	var array6x6 [6][6]int

	for i := 0; i < 6; i++ {
		for j := 0; j < 6; j++ {
			fmt.Scanf("%d", &array6x6[i][j])
		}
	}

	max := math.MinInt32

	for i := 1; i < 5; i++ {
		for j := 1; j < 5; j++ {
			sumHourglass := computeSumHourglass(array6x6, i, j)
			if sumHourglass > max {
				max = sumHourglass
			}
		}
	}

	fmt.Printf("%d", max)
}

func computeSumHourglass(arr [6][6]int, i, j int) (sum int) {
	return arr[i-1][j-1] + arr[i-1][j] + arr[i-1][j+1] +
		arr[i][j] +
		arr[i+1][j-1] + arr[i+1][j] + arr[i+1][j+1]
}

File: data-structures/arrays/2D Array - DS/solution.js
// NOTE to run it
// node solution.js  < input.txt

process.stdin.resume();
process.stdin.setEncoding('ascii');

var input_stdin = "";
var input_stdin_array = "";
var input_currentline = 0;

process.stdin.on('data', function (data) {
    input_stdin += data;
});

process.stdin.on('end', function () {
    input_stdin_array = input_stdin.split("\n");
    main();
});

function readLine() {
    return input_stdin_array[input_currentline++];
}

/////////////// ignore above this line ////////////////////

const getHourglassSum = (array2D, middle) => {
  const { x, y } = middle;

  const value = array2D[y - 1][x - 1] + array2D[y - 1][x] + array2D[y - 1][x + 1]
    + array2D[y][x]
    + array2D[y + 1][x - 1] + array2D[y + 1][x] + array2D[y + 1][x + 1];

  return value;
}

function main() {
  var arr = [];

  for(arr_i = 0; arr_i < 6; arr_i++) {
    arr[arr_i] = readLine().split(' ');
    arr[arr_i] = arr[arr_i].map(Number);
  }

  let maxSum = -1337;

  for (let y = 1 ; y < 5 ; y++) {
    for (let x = 1 ; x < 5 ; x++) {
    const hourGlassValue = getHourglassSum(arr, { x, y });
    maxSum = hourGlassValue > maxSum ? hourGlassValue : maxSum;
    }
  }

  console.log(maxSum);
}
